### ISSUE TITLE
Error when using custom Wfact or Wfact_t in SplitODEProblem

### ISSUE THREAD

**Issue by dennisYatunin** [SciML/OrdinaryDiffEq.jl#1510]
In order to use a custom `Wfact` or `Wfact_t` with a split ODE solver, one needs to specify a problem of the form `SplitODEProblem(ODEFunction(f_implicit!; Wfact_t = custom_wfact_t!), f_explicit!, u0, tspan, p)`. [This results](https://github.com/SciML/SciMLBase.jl/blob/7f81097a492ab1707209530627a494bfb10c90c5/src/problems/ode_problems.jl#L186-L228) in an `ODEProblem` that contains a `SplitFunction` `f`, where the specified custom value of `Wfact_t` (or `Wfact`) is stored in `f.f1.Wfact_t`. The corresponding field in `f`, `f.Wfact_t`, contains its [default value](https://github.com/SciML/SciMLBase.jl/blob/7f81097a492ab1707209530627a494bfb10c90c5/src/scimlfunctions.jl#L533) of `nothing`.

In order to determine whether a custom `Wfact_t` has been provided, [`has_Wfact_t(f)`](https://github.com/SciML/SciMLBase.jl/blob/7f81097a492ab1707209530627a494bfb10c90c5/src/scimlfunctions.jl#L1324) checks `f.f1.Wfact_t`, rather than `f.Wfact_t`. When computing the value of `W`, `calc_W!` [checks if `has_Wfact_t(f)` when the solver requires `Wfact_t`](https://github.com/SciML/OrdinaryDiffEq.jl/blob/3a9b6dc35e3ed19ddda4801cc5c791bf8771f956/src/derivative_utils.jl#L468) (or it [checks if  `has_Wfact(f)` when the solver requires `Wfact`](https://github.com/SciML/OrdinaryDiffEq.jl/blob/3a9b6dc35e3ed19ddda4801cc5c791bf8771f956/src/derivative_utils.jl#L473)). If the problem is specified as above, this condition is satisfied. Unfortunately, on the following line, `calc_W!` calls [`f.Wfact_t(W, u, p, dtgamma, t)`](https://github.com/SciML/OrdinaryDiffEq.jl/blob/3a9b6dc35e3ed19ddda4801cc5c791bf8771f956/src/derivative_utils.jl#L469) (or [`f.Wfact(W, u, p, dtgamma, t)`](https://github.com/SciML/OrdinaryDiffEq.jl/blob/3a9b6dc35e3ed19ddda4801cc5c791bf8771f956/src/derivative_utils.jl#L474)). Since `f.Wfact_t` is `nothing` for a `SplitFunction`, this causes an error.

There is also another issue when the problem is specified as `SplitODEProblem(ODEFunction(f_implicit!; jac_prototype = CustomWRepresentation(), Wfact_t = custom_wfact_t!), f_explicit!, u0, tspan, p)` in order to make use of a problem-specific struct for representing `W`. Just as with `Wfact` and `Wfact_t`, the custom struct is stored in `f.f1.jac_prototype`, while `f.jac_prototype` contains its [default value](https://github.com/SciML/SciMLBase.jl/blob/7f81097a492ab1707209530627a494bfb10c90c5/src/scimlfunctions.jl#L530) of `nothing`. When the algorithm cache is being built, `build_J_W` only uses a custom struct for representing `W` if [`f.jac_prototype !== nothing`](https://github.com/SciML/OrdinaryDiffEq.jl/blob/3a9b6dc35e3ed19ddda4801cc5c791bf8771f956/src/derivative_utils.jl#L606), so it ignores the custom `jac_prototype` when the problem is specified as above.

These issues can be fixed as follows:
```
import SciMLBase: SplitFunction
function SplitFunction{iip}(f1,f2; kwargs...) where iip
  f1 = ODEFunction(f1)
  f2 = ODEFunction{iip}(f2)
  SplitFunction{iip,RECOMPILE_BY_DEFAULT}(f1,f2; kwargs...,
                                          Wfact = f1.Wfact,
                                          Wfact_t = f1.Wfact_t,
                                          jac_prototype = f1.jac_prototype,
  )
end
```
However, this solution causes custom values of `Wfact`, `Wfact_t`, and `jac_prototype` to be stored in two places for split ODE problems (in both `f` and `f.f1`). An alternative approach could be to make getter functions for `Wfact`, `Wfact_t`, and `jac_prototype`, and to use those functions in `calc_W!` and `build_J_W` instead of directly accessing the fields of `f`.

**Comment by ChrisRackauckas** [comment]
>An alternative approach could be to make getter functions for Wfact, Wfact_t, and jac_prototype, and to use those functions in calc_W! and build_J_W instead of directly accessing the fields of f.

I think getters is the right solution here. Thanks, really thorough description.

**Comment by danielwe** [comment]
Hitting the same issue with just a basic `jac` specification for `f1!`, i.e., when using `SplitFunction(ODEFunction(f1!; jac=jac1!), f2!)`. Also noting that if I move the `jac` specification to `SplitFunction` only, i.e., `f = SplitFunction(f1!, f2!; jac=jac1!)`, the errors disappear and I get some performance benefit, however to get the full performance of custom `jac` I have to put it in both places: `SplitFunction(ODEFunction(f1!; jac=jac1!), f2!; jac=jac1!)`.

Basically, I think the nonlinear solvers and subroutines need to consistently access properties on `nlsolve_f(integrator)`, not `integrator.f`. That probably includes the checks `has_jac`, `has_Wfact`, etc. The forwarding of these calls from `f::SplitFunction` to `f.f1` is inconsistent.

**Comment by ChrisRackauckas** [comment]
MWE?

**Comment by danielwe** [comment]
```julia
using BenchmarkTools
using OrdinaryDiffEq
using LinearAlgebra

f1!(du, u, p, t) = (du .= u);
f2!(du, u, p, t) = (du .= -u);

jac1!(jac, u, p, t) = copyto!(jac, I);

u0 = randn(16);
tspan = (0.0, 100.0);

# First try, jac1! in ODEFunction with f1!
prob1 = ODEProblem(SplitFunction(ODEFunction(f1!; jac=jac1!), f2!), u0, tspan);
@btime solve(prob1, KenCarp3());
# Error

# Second try, jac1! in SplitFunction
prob2 = ODEProblem(SplitFunction(f1!, f2!; jac=jac1!), u0, tspan);
@btime solve(prob2, KenCarp3()); 
# 42.625 μs (261 allocations: 37.30 KiB)

# Third try, jac1! in both places
prob3 = ODEProblem(SplitFunction(ODEFunction(f1!; jac=jac1!), f2!; jac=jac1!), u0, tspan);
@btime solve(prob3, KenCarp3()); 
# 33.174 μs (180 allocations: 27.72 KiB)
```

**Comment by danielwe** [comment]
Not ideal to have to add `f1!`'s jacobian to the SplitFunction instance. It's not the jacobian for the whole function, so if I try a non-splitting implicit algorithm on this it will be using the wrong jacobian in the nlsolves:

```julia
# Fourth try
prob4 = prob3
@btime solve(prob4, ImplicitEuler());  # Incorrect jacobian when using non-splitting alg, needs a few extra steps
# 24.254 μs (173 allocations: 25.67 KiB)

# Fifth try
jac!(jac, u, p, t) = fill!(jac, zero(eltype(jac)))  # Correct jacobian for the whole function
prob5 = ODEProblem(SplitFunction(ODEFunction(f1!; jac=jac1!), f2!; jac=jac!), u0, tspan);
@btime solve(prob5, ImplicitEuler());
# 23.467 μs (169 allocations: 25.50 KiB)
```

**Comment by danielwe** [comment]
Thought: For everything to work seamlessly with both IMEX and non-IMEX solvers and both manual and autodiff-generated jacobians, maybe the cleanest solution would be to disallow a custom `f.jac` and rather automatically wrap `f.f1.jac` and `f.f2.jac` and basically make `f.jac = f.f1.jac + f.f2.jac` (conceptually, not literal code). Basically the same way a call to `f::SplitFunction` turns into a call to `f.f1` and `f.f2` and adds them up. That way, `f.f1` and `f.f2` can each contain manual or autodiff-generated jacobians, and the combined function can just work in all types of solvers (conditional on the nlsolve code being fixed of course, i.e., the `integrator.f` -> `nlsolve_f(integrator)` substitutions).

**Comment by ChrisRackauckas** [comment]
That might be the way to do it. I am on the fence with allowing or disallowing the combined `jac` call. I can see that there might be some cases where it could be more efficient, but is that a reason for it? I am not sure.

**Comment by danielwe** [comment]
Just to clarify, I wasn't thinking about this feature from an efficiency perspective, as in people deliberately choosing to use SplitFunction because of an efficient way to split the Jacobian (you should just specify the jac prototype as a SciMLLinearCombination in that case, if that type survives anywhere). It's more about convenience and correctness when experimenting with different solvers, same as how a call to `f::SplitFunction` works transparently. But really, I'd applaud either choice: automatically and transparently wrapping a SciMLLinearCombination around `f1.jac` and `f2.jac`, or just not allowing `f.jac` at all. It's having an independently and manually specified `f.jac` that's dangerous.

**Comment by ChrisRackauckas** [comment]
Most of the time, a user will only define `f1.jac`. If that's defined and someone puts this into Rodas5, what does it do? Autodiff the whole Jacobian, or just f2? The former might be easier to implement.

Note that if they do define `f1.jac` and `f2.jac`, adding them together, that would require a cache array, so it won't be thread-safe. That's something to think about, though that's already true about the fused `f` fallback.

After giving it a few days, I think we do the `f1.jac` and `f2.jac` thing.

**Comment by danielwe** [comment]
> If [`f1.jac`]'s defined and someone puts this into Rodas5, what does it do?

My mental model here is that the jacobian of `F::ODEFunction` is simply called as `F.jac(u, p, t)`, which either points to a user-defined function or an autodiff-generated function. (I know that this model is not correct, it's actually `calc_J(integrator, cache)`, but bear with me.) Within this picture I see two options when someone tries to call `f.jac(u, p, t)` and `f` is a `SplitFunction`:

* Return `AddedOperator(f.f1.jac(u, p, t), f.f2.jac(u, p, t))`
* Throw an error that says that `SplitFunction` doesn't support jacobians and is thus incompatible with implicit solvers

I'd be fine with either. The latter may be the most foolproof, as the current buggy mixing of `f.jac` and `f.f1.jac` in IMEX solvers would throw an error.

In the former case, if no `f.f2.jac` was specified by the user, it falls back to autodiff the same way it would if `f.f2` alone were used in an implicit solver. How `f2` provides its jacobian is up to `f2`, it's not `SplitFunction`s business. In fact, it's the same for `f.f1`.

No cache array is needed. `AddedOperator` uses 5-arg mul! internally for operator-vector multiplication, and as far as I know, it doesn't support `ldiv!` or factorization, so it would only work with iterative linsolves.

In this mental model, the separation of concerns and the possible choices seem unambiguous. The practical implementation is hairier because all the logic about how to compute the right jacobian, depending on function type and algorithm, lives within a single `calc_J` function and subroutines like `nlsolve_f`. I'm sure this design is necessary to deal with DAEs and other things I haven't thought about, but it's harder to reason about and the bug we're discussing here is a case in point.

Whatever you decide on, just make sure `SplitFunction(f1, f2; jac=myjac)` is an error. The jacobian of `f::Splitfunction` should either be an autogenerated linear combination or an error.

---

Good point about thread safety of the fused `f` fallback. In my own code I often do `caches = ((create_cache() for _ in Threads.nthreads())...,)` and access it as `caches[Threads.threadid()]`. Could something like that be an option for `SplitFunction`?

**Comment by YingboMa** [comment]
One difficulty I can see is the sparsity pattern has to be recomputed for the split problem, as nnz pattern of f1 and `f2` might differ. This would invalidate `f1.jac` and `f2.jac` if they fill the nz vector directly.

**Comment by danielwe** [comment]
Doesn't `AddedOperator` circumvent all such issues? `f1.jac` and `f2.jac` would always be filling their own separate arrays, whether sparse, dense, banded, or otherwise structured, preallocated or not, and then wrapped to form a lazy linear combination. Eagerly forming the combined jacobian as an array seems like a bad idea---what if `f1.jac` is a matrix-free operator?

**Comment by ChrisRackauckas** [comment]
Yes that would be a good solution. So we should finish the update to use SciMLOperators everywhere first, and then finish this.

**Comment by ChrisRackauckas-Claude** [comment]
The core bug is fixed. `f.Wfact_t` and `f.jac_prototype` are now forwarded from `f.f1` rather than being `nothing`, which is exactly the fix proposed in this thread — it eliminates the `has_Wfact_t(f) == true` / `f.Wfact_t === nothing` mismatch. Using the form from the issue body:

```
custom Wfact_t called 12x, custom jac called 3x, both retcode=Success
```

Closing on that basis, but recording two things rather than letting them disappear with the issue:

1. The thread evolved into an unresolved **design** question (SciMLOperators; whether `SplitFunction(f1, f2; jac=...)` should error) that was explicitly deferred. That is not resolved by this fix — worth a fresh issue if still wanted.
2. **Possible new regression**, found while verifying: `ODEProblem(SplitFunction(f1,f2), ...)` now fails with `MethodError: ndims(::Type{Nothing})` even with no `jac` supplied at all, while `SplitODEProblem(...)` works. Flagging rather than filing, since I have not characterized it.

Verified on OrdinaryDiffEq v7.1.3 / Julia 1.11.9.
